### ISSUE TITLE
Searching sharee with displayname

### ISSUE THREAD

**Issue by jesmrec** [owncloud/ocis#547]
Only posible to get sharees by using the username:

Username: `einstein`
Display Name: `Albert Einstein`

results are fetched only matching `einstein`
`Albert`, `Einstein`, `Ein`, `Albe` do not match and response is empty (always forgetting the case sensitiveness)

oC10: matches with both.

Backend: EOS storage via via https://github.com/owncloud-docker/compose-playground/blob/master/examples/hetzner-deploy/make_ocis_eos_compose_test.sh
Version: v1.0.0-rc1
Client: iOS App v11.4 over iOS13

CC @pmaier1

**Comment by refs** [comment]
I added the EOS storage tag but most likely is related with the userprovider.

**Comment by micbar** [comment]
re tested:

Created User

Username: hans
Display Name: Peter Dampf
E-Mail: pd@example.com

We can find:

`hans`
`Peter`
`pd@`

We cannot find:

`Hans` (should not be case sensitive)
`ans`(middle of the username)
`Dampf` (End of the string Displayname)
 `example.com` (End of the email address)

## Expectation

at least
`Hans` (username search should not be case sensitive)
`Dampf` (End or middle of the string Displayname)

Should work.

**Comment by rhafer** [comment]
> ## Expectation
> 
> at least
> `Hans` (username search should not be case sensitive)

This is caused by a bug in idm. Fix is on its way.

> `Dampf` (End or middle of the string Displayname)

This is currently not working "on purpose". At some point we agreed on limiting search to prefix searches. As doing full substring searches will get pretty expensive easily. We can of course change that (just need to adjust the filter https://github.com/cs3org/reva/blob/75f2c2f836578f8280749569f7f01a3e7d2b5939/pkg/utils/ldap/identity.go#L480).

@micbar If you still want it. Should we maybe make that configurable?

**Comment by rhafer** [comment]
> > ## Expectation
> > at least
> > `Hans` (username search should not be case sensitive)

This part should be fixed with  https://github.com/libregraph/idm/pull/75

**Comment by micbar** [comment]
> @micbar If you still want it. Should we maybe make that configurable?

That would be awesome, same for the character length.

**Comment by rhafer** [comment]
> same for the character length.

@micbar The API does not enforce any limit on the minimum (or maximum) length of the search phrase. Can you elaborate on what you are referring to?

**Comment by micbar** [comment]
Currently the system starts searching for users after you enter 3 characters. This is fine as a default.

Some use cases and the test suite would like to use at least 2 chars to start the user enumeration. I would like to have that configurable. Is that possible?

**Comment by phil-davis** [comment]
For example, oC10 has:
```
'user.search_min_length' => 2,
```

So the default on oC10 is to start searching for matches after 2 characters are typed in.

For test scenarios, it will be easier if we can get oCIS to start with this set to 2 characters (either change the oCIS default from 3 to 2, or have a way that we can configure that when we start oCIS in CI - I don't really care)

There are some scenarios that change the setting on-the-fly and check that different values work. Tht is out-of-scope for oCIS at the moment - we don't have a general way to modify system-wide settings on-the-fly in oCIS, so those sort of test scenarios will have to be re-engineered anyway if there comes a time when we actually want to test such things.

**Comment by pmaier1** [comment]
Configurable would be great, default like oC10 (`2`)

**Comment by SagarGi** [comment]
This PR got merged https://github.com/owncloud/ocis/pull/4229 fixing this issue and needs a follow up for the expected to failure.

https://github.com/owncloud/ocis/blob/176dd321b1f8b85a45440119541cc3be17fe0125/tests/acceptance/expected-failures-API-on-OCIS-storage.md?plain=1#L409-L421

TODO QA-team
- [ ] review tests linked to this issue
- [ ] remove them from expected to fail if they pass or create other issue if necessary

**Comment by SwikritiT** [comment]
The tests are failing because they search for the `exact` match for a searched string but this field is always empty for `OCIS`
```
<exact>
      <users/>
      <groups/>
      <remotes/>
    </exact>

```

For instance if I create a multiple variations of user `einstein`  in `OC10` and search it with curl 
```console
curl -vk -uadmin:admin http://localhost/core/ocs/v1.php/apps/files_sharing/api/v1/sharees\?search\=einstein\&itemType\=file -H"OCS-APIREQUEST:1" | xmllint --format -
```
It responds with 
```console
<?xml version="1.0"?>
<ocs>
  <meta>
    <status>ok</status>
    <statuscode>100</statuscode>
    <message>OK</message>
    <totalitems/>
    <itemsperpage/>
  </meta>
  <data>
    <exact>
      <users>
        <element>
          <label>einstein</label>
          <value>
            <shareType>0</shareType>
            <shareWith>einstein</shareWith>
            <userType>0</userType>
          </value>
        </element>
      </users>
      <groups/>
      <remotes/>
    </exact>
    <users>
      <element>
        <label>Albert Einstein</label>
        <value>
          <shareType>0</shareType>
          <shareWith>Albert Einstein</shareWith>
          <userType>0</userType>
        </value>
      </element>
      <element>
        <label>einstein-einstein</label>
        <value>
          <shareType>0</shareType>
          <shareWith>einstein-einstein</shareWith>
          <userType>0</userType>
        </value>
      </element>
      <element>
        <label>einstein1</label>
        <value>
          <shareType>0</shareType>
          <shareWith>einstein1</shareWith>
          <userType>0</userType>
        </value>
      </element>
    </users>
    <groups/>
    <remotes/>
  </data>
</ocs>

```
Inside `<exact><users>` there's `exact` match of the string 

But in case of `OCIS` 
```
curl -vk -uadmin:admin https://host.docker.internal:9200/ocs/v1.php/apps/files_sharing/api/v1/sharees\?search\=einstein\&itemType\=file -H"OCS-APIREQUEST:1" | xmllint --format - 
```

```
<?xml version="1.0" encoding="UTF-8"?>
<ocs>
  <meta>
    <status>ok</status>
    <statuscode>100</statuscode>
    <message>OK</message>
  </meta>
  <data>
    <exact>
      <users/>
      <groups/>
      <remotes/>
    </exact>
    <users>
      <element>
        <label>Albert Einstein</label>
        <value>
          <shareType>0</shareType>
          <shareWith>einstein</shareWith>
          <shareWithAdditionalInfo>einstein@example.org</shareWithAdditionalInfo>
        </value>
      </element>
      <element>
        <label>Einstein Albert</label>
        <value>
          <shareType>0</shareType>
          <shareWith>einstein-einstein</shareWith>
          <shareWithAdditionalInfo>a@c.com</shareWithAdditionalInfo>
        </value>
      </element>
      <element>
        <label>Einstein</label>
        <value>
          <shareType>0</shareType>
          <shareWith>einstein1</shareWith>
          <shareWithAdditionalInfo>a@b.com</shareWithAdditionalInfo>
        </value>
      </element>
    </users>
    <groups/>
    <remotes/>
  </data>
</ocs>
```
The `exact` match is rather given in `<users>` 
```
curl -vk -uadmin:admin https://host.docker.internal:9200/ocs/v1.php/apps/files_sharing/api/v1/sharees\?search\=einstein-einstein\&itemType\=file -H"OCS-APIREQUEST:1" | xmllint --format -
```
```
<?xml version="1.0" encoding="UTF-8"?>
<ocs>
  <meta>
    <status>ok</status>
    <statuscode>100</statuscode>
    <message>OK</message>
  </meta>
  <data>
    <exact>
      <users/>
      <groups/>
      <remotes/>
    </exact>
    <users>
      <element>
        <label>Einstein Albert</label>
        <value>
          <shareType>0</shareType>
          <shareWith>einstein-einstein</shareWith>
          <shareWithAdditionalInfo>a@c.com</shareWithAdditionalInfo>
        </value>
      </element>
    </users>
    <groups/>
    <remotes/>
  </data>
</ocs>
```

The tests are looking inside `exact` and they don't find the expected value there.

Is this expected behaviour?

**Comment by rhafer** [comment]
@SwikritiT thanks for looking into that. I think this deserves a separate issue though, would you mind creating a new one for this?

I think the original bug reported here (which was about broken substring and case-sensitiveness) was solved.

It seems the new issue is basically caused by an incomplete implementation of the ocs sharing api in ocis/reva. It always returns an empty `exact` list and stuffs all the results into the top level `users` list  (https://github.com/cs3org/reva/blob/master/internal/http/services/owncloud/ocs/handlers/apps/sharing/sharees/sharees.go#L91)

**Comment by SwikritiT** [comment]
> @SwikritiT thanks for looking into that. I think this deserves a separate issue though, would you mind creating a new one for this?
> 
> I think the original bug reported here (which was about broken substring and case-sensitiveness) was solved.
> 
> It seems the new issue is basically caused by an incomplete implementation of the ocs sharing api in ocis/reva. It always returns an empty `exact` list and stuffs all the results into the top level `users` list (https://github.com/cs3org/reva/blob/master/internal/http/services/owncloud/ocs/handlers/apps/sharing/sharees/sharees.go#L91)

Okay I'll create a separate issue and close this one

**Comment by SwikritiT** [comment]
@rhafer new issue created here https://github.com/owncloud/ocis/issues/4265

**Comment by SwikritiT** [comment]
> This PR got merged #4229 fixing this issue and needs a follow-up for the expected to fail.
> 
> https://github.com/owncloud/ocis/blob/176dd321b1f8b85a45440119541cc3be17fe0125/tests/acceptance/expected-failures-API-on-OCIS-storage.md?plain=1#L409-L421
> 
> TODO QA-team
> 
> * [x]  review tests linked to this issue
> * [x]  remove them from expected to fail if they pass or create another issue if necessary -> new issue created here [ocs sharing api always returns an empty exact list while searching for a sharee #4265](https://github.com/owncloud/ocis/issues/4265)

Closing this issue.
